presto-main/src/main/java/com/facebook/presto/execution/buffer/ArbitraryOutputBuffer.java

```diff
@@ -219,10 +226,23 @@ public void enqueue(List<SerializedPage> pages)
         long rowCount = pages.stream().mapToLong(SerializedPage::getPositionCount).sum();
         totalRowsAdded.addAndGet(rowCount);
         totalPagesAdded.addAndGet(pages.size());
+        outstandingPagePerLifespan.computeIfAbsent(lifespan, ignore -> new AtomicLong()).addAndGet(pages.size());
 
         // create page reference counts with an initial single reference
         List<SerializedPageReference> serializedPageReferences = pages.stream()
-                .map(pageSplit -> new SerializedPageReference(pageSplit, 1, () -> memoryManager.updateMemoryUsage(-pageSplit.getRetainedSizeInBytes())))
+                .map(pageSplit -> new SerializedPageReference(
+                        pageSplit,
+                        1,
+                        () -> {
+                            long oustandingPage = outstandingPagePerLifespan.get(lifespan).decrementAndGet();
+                            Consumer<Lifespan> callback = lifespanFinishedCallback;
+                            if (oustandingPage == 0 && callback != null
+                                    && noMorePagesForLifespan.contains(lifespan)) {
```

Comment:
No need to have a newline in this if condition here. Also reads well if all conditions are on the same line.